types/mysql__xdevapi/index.d.ts

```diff
@@ -0,0 +1,493 @@
+}
+
+export interface DatabaseObject {
+	getSession(): Session;
+}
+
+export interface Grouping {
+	groupBy(...GroupByExprStr: string[]): Grouping;
+	groupBy(GroupByExprStr: string | ReadonlyArray<string>): Grouping;
+
+	having(expr: SearchConditionStr): Grouping;
+}
+
+export interface Limiting {
+	limit(count: number, offset?: number): Limiting;
+
+	offset(value: number): Limiting;
+}
+
+export interface Locking {
+	lockExclusive(mode?: LockContention): Locking;
+
+	lockShared(mode?: LockContention): Locking;
+}
+
+export interface Statement {
+	addArgs(args: ReadonlyArray<any>): Statement;
+
+	getArgs(): any[];
+
+	getNamespace(): string;
+
+	getRawStatement(): string;
+
+	getSession(): Session;
+}
+
+export interface TableFiltering {
+	where(criteria: string): TableFiltering;
+}
+
+export interface TableOrdering {
+	orderBy(SortExprStr: string | ReadonlyArray<string>): TableOrdering;
+}
+
+export class Collection<T> implements DatabaseObject {
+	add(expr: T): CollectionAdd<T>;
+
+	addOrReplaceOne(id: string, data: T): Promise<Result>;
+
+	count(): Promise<number>;
+
+	createIndex(
+		name: string,
+		constraint: IndexDefinition
+	): Promise<boolean>;
+
+	dropIndex(name: string): Promise<boolean>;
+
+	existsInDatabase(): Promise<boolean>;
+
+	find(expr: SearchConditionStr): CollectionFind<T>;
+
+	getName(): string;
+
+	getOne(id: string): T;
+
+	getSchema(): Schema;
+
+	inspect(): { schema: string; collection: string };
+
+	modify(expr: SearchConditionStr): CollectionModify<T>;
+
+	remove(expr: SearchConditionStr): CollectionRemove<T>;
+
+	removeOne(id: string): Promise<Result>;
+
+	replaceOne(id: string, data: T): Promise<Result>;
+
+	getSession(): Session;
+}
+
+export class CollectionAdd<T> {
+	add(input: T | ReadonlyArray<T>): CollectionAdd<T>;
+
+	execute(): Promise<Result>;
+}
+
+export class CollectionFind<T>
+	implements Binding, CollectionOrdering, Grouping, Limiting, Locking {
+	execute(callback: (fields: T) => void): Promise<Result>;
+
+	fields(projections: ReadonlyArray<string> | string): CollectionFind<T>;
+
+	bind(parameter: any): CollectionFind<T>;
+	bind(parameter: string, value: any): CollectionFind<T>;
+
+	sort(...SortExprStr: string[]): CollectionFind<T>;
+	sort(SortExprStr: string | ReadonlyArray<string>): CollectionFind<T>;
+
+	groupBy(...GroupByExprStr: string[]): CollectionFind<T>;
+	groupBy(GroupByExprStr: string | ReadonlyArray<string>): CollectionFind<T>;
+
+	having(expr: SearchConditionStr): CollectionFind<T>;
+
+	limit(count: number, offset?: number): CollectionFind<T>;
+
+	offset(value: number): CollectionFind<T>;
+
+	lockExclusive(mode: LockContention): CollectionFind<T>;
+
+	lockShared(mode: LockContention): CollectionFind<T>;
+}
+
+export class CollectionModify<T> implements Binding, CollectionOrdering, Limiting {
+	arrayAppend<
+		K extends keyof T = keyof T,
+		// tslint:disable-next-line:no-unnecessary-generics
+		J extends keyof T[K] = keyof T[K]
```

Comment:
I don't think you need these defaults here (e.g. `= keyof T`).
In fact, you don't even need `J`; you can just index with `keyof T[K]`.